frontend/english-keyset.ts

```diff
@@ -0,0 +1,339 @@
+        return (e: KeyboardEvent) => { console.log(key_str + " Key Up") };
+    }
+
+    private getMouseDownFunction(key_str: string) {
+
+        let customMouseDownFn = () => {
+
+            if (key_str.toLowerCase() in ["shift", "ctrl", "alt"]) {
+                let curKeyState = Boolean(this.containingKeyboard.states[key_str.toLowerCase()]);
+                this.containingKeyboard.states[key_str.toLowerCase()] = !curKeyState;
+            }
+            else {
+
+                if (this.containingKeyboard.targetElement instanceof HTMLInputElement ||
+                    this.containingKeyboard.targetElement instanceof HTMLTextAreaElement) {
+
+                    let curValue = this.containingKeyboard.targetElement.value;
+                    let modifiedValue = getModifiedValueOnKeyPress(curValue, key_str, this.containingKeyboard.states);
+                    this.containingKeyboard.targetElement.value = modifiedValue;
+
+                }
+                else if (typeof (this.containingKeyboard.targetElement) === "string") {
+
+                    let curValue = this.containingKeyboard.targetElement;
+                    let modifiedValue = getModifiedValueOnKeyPress(curValue, key_str, this.containingKeyboard.states);
+                    this.containingKeyboard.targetElement = modifiedValue;
+                }
+            }
+
+            if (this.containingKeyboard.targetElement instanceof HTMLElement) {
+                this.containingKeyboard.targetElement.dispatchEvent(new Event('input'));  // notify value update
+                this.containingKeyboard.targetElement.dispatchEvent(new KeyboardEvent('keydown', {
+                    key: key_str,
+                    code: key_str,
+                    ctrlKey: Boolean(this.containingKeyboard.states['ctrl']),
+                    shiftKey: Boolean(this.containingKeyboard.states['shift']),
+                    altKey: Boolean(this.containingKeyboard.states['alt']),
+                }));
+            }
+        };
+
+        return () => {
+            customMouseDownFn();
+        };
+    }
+
+    private getMouseUpFunction(key_str: string) {
+        return (e: MouseEvent) => { };
+    }
+
+    private getKeySpan(key: string): number {
+        return key === "Space"
+            ? 14
+            : key === "Shift"
+                ? 5
+                : key === "Delete" ||
+                    key === "Enter" ||
+                    key === "Ctrl" ||
+                    key === "Alt" ||
+                    key === "Caps"
+                    ? 4
+                    : key === "Tab" || key === "\\"
+                        ? 3
+                        : 2;
+    }
+
+    private isSpecialKey(key: string): boolean {
+        return [
+            "Delete",
+            "Enter",
+            "Shift",
+            "Caps",
+            "Tab",
+            "Ctrl",
+            "Alt",
+            "Win",
+            "Space",
+        ].includes(key);
+    }
```

Comment:
wouldn't all this code get repeated for other languages as well?
can you come up with a way to generalize it?

Reply:
Depending on the other language keyboards made, a lot of it MIGHT be repeated, so generalizing it between other languages was the target in next iteration.

But there's also the case that the keysets can change and multiple types of keyboards for the same languages can be made.

For example,

There can be a keyboard which resembles the MAC Keyboard, with the "Command" key and everything.
There can be a keyboard without the capsLock key(like we have in smartphones)

So if we're looking to generalize this functionality, maybe we can make a "SimpleQWERTYKeyset" base class or something and move this code there, and then use that base class to make our "simple" keyboards, but I wont recommend forcing these keys on the entire Keyboard like the previous iteration was.